### PR TITLE
[BEAM-4420] KafkaIOIT metrics collection

### PR DESCRIPTION
**Pull request by lgajowy**
Added "TimeMonitors" and collect read, write and run time. Some minor refactoring included (in IOITMetrics.java class).

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/)
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_Verify/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python3_Verify/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/) <br> [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_PVR_Flink_Cron/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/)

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website
--- | --- | --- | --- | ---
Non-portable | [Build Status](https://builds.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) 
Portable | --- | [Build Status](https://builds.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

### CODE REVIEW

**Comment by lgajowy** [comment]
@aromanenko-dev could you take a look?

File: sdks/java/io/kafka/src/test/java/org/apache/beam/sdk/io/kafka/KafkaIOIT.java [hunk 1]
```diff
@@ -79,22 +95,53 @@ public static void setup() throws IOException {
...
+
+    Set<NamedTestResult> metrics = readMetrics(writeResult, readResult);
+    IOITMetrics.publish(
+        testId, timestamp, options.getBigQueryDataset(), options.getBigQueryTable(), metrics);
+  }
+
+  private Set<NamedTestResult> readMetrics(PipelineResult writeResult, PipelineResult readResult) {
+    Function<MetricsReader, NamedTestResult> writeTimeSupplier =
+        reader -> {
+          long start = reader.getStartTimeMetric("write_time");
```
Comment (aromanenko-dev) [inline sdks/java/io/kafka/src/test/java/org/apache/beam/sdk/io/kafka/KafkaIOIT.java]: Please, extract "write_time" as a constant value and use this value over the test instead.

File: sdks/java/io/kafka/src/test/java/org/apache/beam/sdk/io/kafka/KafkaIOIT.java [hunk 2]
```diff
@@ -79,22 +95,53 @@ public static void setup() throws IOException {
...
+          long end = reader.getEndTimeMetric("write_time");
+          return NamedTestResult.create(testId, timestamp, "write_time", (end - start) / 1e3);
+        };
+
+    Function<MetricsReader, NamedTestResult> readTimeSupplier =
+        reader -> {
+          long start = reader.getStartTimeMetric("read_time");
```
Comment (aromanenko-dev) [inline sdks/java/io/kafka/src/test/java/org/apache/beam/sdk/io/kafka/KafkaIOIT.java]: Please, extract "read_time" as a constant value and use this value over the test instead.

File: sdks/java/io/kafka/src/test/java/org/apache/beam/sdk/io/kafka/KafkaIOIT.java [hunk 3]
```diff
@@ -79,22 +95,53 @@ public static void setup() throws IOException {
...
+  private Set<NamedTestResult> readMetrics(PipelineResult writeResult, PipelineResult readResult) {
+    Function<MetricsReader, NamedTestResult> writeTimeSupplier =
+        reader -> {
+          long start = reader.getStartTimeMetric("write_time");
+          long end = reader.getEndTimeMetric("write_time");
+          return NamedTestResult.create(testId, timestamp, "write_time", (end - start) / 1e3);
+        };
+
+    Function<MetricsReader, NamedTestResult> readTimeSupplier =
+        reader -> {
```
Comment (aromanenko-dev) [inline sdks/java/io/kafka/src/test/java/org/apache/beam/sdk/io/kafka/KafkaIOIT.java]: Looks like the code of `readTimeSupplier` is very similar to code of `writeTimeSupplier`. Can we have only one parameterised implementation for both cases?

**Comment by aromanenko-dev** [comment]
Also, do you think it would be better to squash commits into one?

**Comment by lgajowy** [comment]
Thanks! I posted the fixes. Could you take a look again?

I also squashed the Kafka part but left out the IOITMetrics refactoring as a separate commit. I think it's a separate concern, more general one that we didn't have a way to publish already collected `NamedTestResult`s. For order and clean git history, I'd like to keep it separate unless you have something against.

**Comment by aromanenko-dev** [comment]
Run Python_PVR_Flink PreCommit

**Comment by aromanenko-dev** [comment]
Run JavaPortabilityApi PreCommit

**Comment by lgajowy** [comment]
Thanks!

**Pull request merged**